compiler/rustc_span/src/def_id.rs

```diff
@@ -171,6 +171,12 @@ impl StableCrateId {
         // linking against a library of the same name, if this is an executable.
         hasher.write(if is_exe { b"exe" } else { b"lib" });
 
+        // Also incorporate the rustc version. Otherwise, with -Zsymbol-mangling-version=v0
+        // and no -Cmetadata, symbols from the same crate compiled with different versions of
+        // rustc are named the same.
+        let rustc_version = option_env!("CFG_VERSION").unwrap_or("unknown version").as_bytes();
```

Comment:
I think you will need to copy the implementation of `rustc_incremental::persist::file_format::rustc_version()` here for `RUSTC_FORCE_INCR_COMP_ARTIFACT_HEADER` to work. Or you could completely move it into this crate and use it here.